test/test_tensorflow.py

```diff
@@ -1746,6 +1746,336 @@ def test_horovod_alltoall_equal_split_gpu(self):
+                    "hvd.alltoall collected wrong number of values")
+
+
+    def test_horovod_alltoall_zero_splits_cpu(self):
+        """Test alltoall with some ranks not participating / splits set to zero."""
+        hvd.init()
+
+        if hvd.size() == 1:
+            self.skipTest("Only one worker available")
+
+        active_ranks = range(0, hvd.size() // 2)
+        silent_ranks = range(hvd.size() // 2, hvd.size())
+
+        active_splits = [1 if r in active_ranks else 0 for r in range(hvd.size())]
+        active_shape = [sum(active_splits), 4]
+        silent_splits = [0] * hvd.size()
+        silent_shape = [0, 4]
+
+        with tf.device("/cpu:0"):
+            if hvd.rank() in active_ranks:
+                source_tensor = tf.fill(active_shape, value=tf.cast(hvd.rank(), tf.int32))
+                splits = tf.convert_to_tensor(active_splits)
+            else:
+                source_tensor = tf.fill(silent_shape, value=tf.cast(hvd.rank(), tf.int32))
+                splits = tf.convert_to_tensor(silent_splits)
+            collected = hvd.alltoall(source_tensor, splits, name="alltoall_zero_splits")
+            result = self.evaluate(collected)
+
+        print(hvd.rank(), "result.shape", result.shape)
+        print(hvd.rank(), "result", result)
+        if hvd.rank() in active_ranks:
+            expected_result_shape = active_shape
+        else:
+            expected_result_shape = silent_shape
+        self.assertSequenceEqual(result.shape, expected_result_shape)
+        if hvd.rank() in active_ranks:
+            for r_idx, r in enumerate(active_ranks):
+                self.assertTrue(np.all(result[r_idx, ...] == r))
+        else:
+            self.assertLen(result, 0)
+
+    def test_horovod_alltoall_zero_splits_gpu(self):
+        """Test alltoall with some ranks not participating / splits set to zero."""
+        # ncclCommInitRank failed: invalid usage
+        hvd.init()
+
+        if hvd.size() == 1:
+            self.skipTest("Only one worker available")
+
+        # Only do this test if there are GPUs available.
+        if not tf.test.is_gpu_available(cuda_only=True):
+            self.skipTest("No GPUs available")
+
+        if int(os.environ.get('HOROVOD_MIXED_INSTALL', 0)):
+            # Skip if compiled with CUDA but without HOROVOD_GPU_OPERATIONS.
+            self.skipTest("Not compiled with HOROVOD_GPU_OPERATIONS")
+
+        # This test does not apply if NCCL version < 2.7.0
+        if hvd.nccl_built() and hvd.nccl_built() < 2700:
+            self.skipTest("NCCL-based Alltoall requires NCCL version >= 2.7.0.")
+
+        active_ranks = range(0, hvd.size() // 2)
+        silent_ranks = range(hvd.size() // 2, hvd.size())
+
+        active_splits = [1 if r in active_ranks else 0 for r in range(hvd.size())]
+        active_shape = [sum(active_splits), 4]
+        silent_splits = [0] * hvd.size()
+        silent_shape = [0, 4]
+
+        with tf.device("/gpu:0"):
```

Comment:
This line is causing the `ncclCommInitRank` failure as all ranks are attempting to initialized NCCL with the same GPU. This line should be `with tf.device("/gpu:%s" % hvd.local_rank()):`

Reply:
Thanks for catching that! And sorry for the confusion!